### DISCUSSION TITLE
Add GitHub actions to validate the serialized files

### DISCUSSION THREAD

**Discussion by lecoqlibre** [datafoodconsortium/taxonomies#26] [Ideas]
Like we said in our last ontology meetings, we should have a way to check whether or not a file is valid before we publish it.

We came with the idea to set up GitHub actions which would be triggered when new commits are pushes to the repo. These actions would test and check some constraints are validated.

Do we need this for both JSON-LD and RDF files?

Possible checks:
- Check naming conventions (see https://github.com/datafoodconsortium/taxonomies/discussions/13)
- Check SKOS broaders and narrowers are valid (see https://github.com/datafoodconsortium/connector-typescript/issues/5#issuecomment-1940821711)
- Check translations are present? Maybe we should have a tool to know what need to be translated. We should look if this can be achieved within VocBench.w

Any comments @RaggedStaff @Alcoz ?

**Comment by RaggedStaff** [comment]
I think some of this might be achieved with [VOCbench's Integrity Constraint Violation checks](https://vocbench.datafoodconsortium.org/vocbench3/#/Icv) (mainly the check SKOS broaders/narrowers/dangling concepts). That said, I don't know if we can automate & I'd prefer we have automated checks in place for every PR, so I think it would be best if they happen on GitHub.

I think we do need to check both JSON-LD & RDF files. 

I agree with all the checks you propose. 

I would also like it if we could have a check to ensure the contents of the file matches the file name (e.g. measures.json contains an `id` of `Measures`, or the rdf files have an xmlns that matches the filename).

**Comment by RaggedStaff** [comment]
On URI naming convention checks:  we need to enforce the following regex's:

camelCase (for properties)  : `"[a-z]+(?:[A-Z0-9]+[a-z0-9]+[A-Za-z0-9]*)*"`
PascalCase (for classes) :  `"(?:[A-Z][a-z0-9]+)(?:[A-Z]+[a-z0-9]*)*"`

On translations: I think for now we want to ensure all concepts have either an English OR French label, does that sound ok? Or should we enforce English AND French for everything  ?

@lecoqlibre Do you have any ideas for the narrower/broader parsing? Would we need to use something that can parse SKOS (like the RDF4J library) or could we use some "simple" awk parsing of the text ? :thinking:

**Reply by RaggedStaff** [reply]
Also just discussing with @Alcoz about SKOS:Concepts. We propose camelCase for these. @lecoqlibre do you agree ?

**Comment by RaggedStaff** [comment]
Was looking at [Super Linter](https://github.com/marketplace/actions/super-linter), but it doesn't support RDF/JSON (or any structured data, except bare XML :confounded: ). 

The [Structured Data Linter](http://linter.structured-data.org/) could be of some use there, but not sure how much if we're outputting from VOCbench. :confused: 

We might also make use of the [W3C RDF Parser](https://github.com/RDFLib/pyrdfa3)

I also came across [this](https://github.com/skohub-io/skohub-vocabs) in my travels... not sure how useful it is atm, but just noting it here.

@lecoqlibre - are you still planning to look into this ? If so, do you have any idea when you might have some bandwidth to do that ?

**Comment by RaggedStaff** [comment]
Could we use the the SHACL validatior from https://github.com/skohub-io/skohub-vocabs to drive this.

@lecoqlibre could you estimate this in an issue ?
